scripts/runcode.jl

```diff
@@ -248,10 +248,286 @@ function chamber(
+- `log_vfr`: Magma recharge rate in km³/yr calculated as 10^(`log_vfr`).
+- `depth`: Depth of the magma chamber in meters.
+- `output_dirname`(optional): Name of the output directory. Defaults to current timestamp.
+
+## Returns
+A `DataFrame` containing the solution with columns:
+- `time`: Simulation timestamps in seconds.
+- `P+dP`: Pressure in Pa.
+- `T`: Temperature in K.
+- `eps_g`: Gas volume fraction.
+- `V`: Volume of the magma chamber in m³.
+- `rho_m`: Density of the melt in kg/m³.
+- `rho_x`: Density of magma crystal in kg/m³.
+- `X_CO2`: Mole fraction of CO2 in the gas.
+- `total_mass`: Total mass of magma chamber in kg.
+- `total_mass_H2O`: Total mass of water in the magma in kg.
+- `total_mass_CO2`: Total mass of CO₂ in the magma in kg.
+
+## Outputs
+A directory named after `output_dirname` or the default value, containing the following files:
+- `out.csv`: a CSV file containing the solution columns listed above.
+- `eruptions.csv`, A CSV file containing the datas of eruptions with the following columns: time_of_eruption (sec), duration_of_eruption (sec), mass_erupted (kg) and volume_erupted (km³).
+- Figures for P+dP(t), T(t), eps_g(t), V(t), X_CO2(t), total_mass(t).
+
+## References
+- W. Degruyter and C. Huber (2014). A model for eruption frequency of upper crustal silicic magma chambers. Earth Planet. Sci. Lett. https://doi.org/10.1016/j.epsl.2014.06.047
+
+## Examples
+```
+# Run a simulation with silicic magma chamber
+julia> composition = Silicic()
+julia> end_time = 3e9
+julia> log_volume_km3 = 0.2
+julia> InitialConc_H2O = 0.04
+julia> InitialConc_CO2 = 0.001
+julia> log_vfr = -3.3
+julia> depth = 8e3
+
+julia> chamber(composition, end_time, log_volume_km3, InitialConc_H2O, InitialConc_CO2, log_vfr, depth)
+
+# Run a simulation with mafic magma chamber, with custom directory name "MyDirname"
+julia> composition = Mafic()
+julia> end_time = 3e9
+julia> log_volume_km3 = 0.2
+julia> InitialConc_H2O = 0.01
+julia> InitialConc_CO2 = 0.001
+julia> log_vfr = -3.3
+julia> depth = 8e3
+julia> output_dirname = "MyDirname"
+
+julia> chamber(composition, end_time, log_volume_km3, InitialConc_H2O, InitialConc_CO2, log_vfr, depth, output_dirname)
+```
+"""
+function chamber(
+    composition::Union{Silicic,Mafic},
+    end_time::Float64,
+    log_volume_km3_vector::Union{Float64,Vector{Float64}},
+    InitialConc_H2O_vector::Union{Float64,Vector{Float64}},
+    InitialConc_CO2_vector::Union{Float64,Vector{Float64}},
+    log_vfr_vector::Union{Float64,Vector{Float64}},
+    depth_vector::Union{Float64,Vector{Float64}},
+    output_dirname::String=get_timestamp(),
+    ;
+    method::String="CVODE_BDF",
+    rheol::String="old",
+)::Nothing
+    datetime = get_timestamp()
+    composition_str = string(typeof(composition))
+    path0 = joinpath(pwd(), output_dirname)
+    mkdir(path0)
+    to = TimerOutput()
+    @timeit to "chamber" begin
+        for log_volume_km3 in vector_uniq(log_volume_km3_vector)
```

Comment:
You could also make the base function _memoized_ to avoid evaluating it again with already evaluated arguments.

Try this on the Julia shell

```julia
julia> using Memoize

julia> @memoize function foo(a)
           @info "Start time $(now())"
           println("Running...")
           sleep(5)
           @info "End time   $(now())"
           return a
       end
foo (generic function with 1 method)

julia> foo(1) # first call
[ Info: Start time 2023-07-13T12:07:56.285
Running...
[ Info: End time   2023-07-13T12:08:01.294
1

julia> foo(1) # second call returns instantly
1
```

Reply:
It's a useful tips! I made the `chamber` function memoized:
https://github.com/brown-ccv/Chamber.jl/blob/2ff308da3875713206be7b41955dd825a99fe095/scripts/runcode.jl#L68